packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,334 @@
+
+type Subscription<T extends IDecodedMessage> = {
+  decoders: IDecoder<T>[];
+  callback: Callback<T>;
+  pubSubTopic: string;
+};
+
+const FilterV2Codecs = {
+  SUBSCRIBE: "/vac/waku/filter-subscribe/2.0.0-beta1",
+  PUSH: "/vac/waku/filter-push/2.0.0-beta1",
+};
+
+/**
+ * Implements client side of the [Waku v2 Filter protocol](https://rfc.vac.dev/spec/12/).
+ *
+ * Note this currently only works in NodeJS when the Waku node is listening on a port, see:
+ * - https://github.com/status-im/go-waku/issues/245
+ * - https://github.com/status-im/nwaku/issues/948
+ */
+class FilterV2 extends BaseProtocol implements IFilterV2 {
+  options: ProtocolCreateOptions;
+  private subscriptions: Map<RequestID, unknown>;
+
+  constructor(public libp2p: Libp2p, options?: ProtocolCreateOptions) {
+    super(
+      FilterV2Codecs.SUBSCRIBE,
+      libp2p.peerStore,
+      libp2p.getConnections.bind(libp2p)
+    );
+
+    this.libp2p.handle(FilterV2Codecs.PUSH, this.onRequest.bind(this));
+
+    this.options = options ?? {};
+    this.subscriptions = new Map();
+  }
+
+  /**
+   * @param decoders Decoder or array of Decoders to use to decode messages, it also specifies the content topics.
+   * @param callback A function that will be called on each message returned by the filter.
+   * @param opts The FilterSubscriptionOpts used to narrow which messages are returned, and which peer to connect to.
+   * @returns Unsubscribe function that can be used to end the subscription.
+   */
+  async subscribe<T extends IDecodedMessage>(
+    decoders: IDecoder<T> | IDecoder<T>[],
+    callback: Callback<T>,
+    opts?: ProtocolOptions
+  ): Promise<UnsubscribeFunction> {
+    const decodersArray = Array.isArray(decoders) ? decoders : [decoders];
+    const { pubSubTopic = DefaultPubSubTopic } = this.options;
+
+    const contentTopics = Array.from(groupByContentTopic(decodersArray).keys());
+
+    const request = FilterSubscribeRpc.createSubscribeRequest(
+      pubSubTopic,
+      contentTopics
+    );
+
+    const { requestId } = request;
+
+    const peer = await this.getPeer(opts?.peerId);
+    const stream = await this.newStream(peer);
+
+    try {
+      const res = await pipe(
+        [request.encode()],
+        lp.encode(),
+        stream,
+        lp.decode(),
+        async (source) => await all(source)
+      );
+
+      const { statusCode, requestId } = FilterSubscribeResponse.decode(
+        res[0].slice()
+      );
+
+      if (statusCode < 200 || statusCode >= 300) {
+        throw new Error(
+          `Filter subscribe request ${requestId} failed with status code ${statusCode}`
+        );
+      }
+
+      log(
+        "Subscribed to peer ",
+        peer.id.toString(),
+        "for content topics",
+        contentTopics
+      );
+    } catch (e) {
+      log(
+        "Error subscribing to peer ",
+        peer.id.toString(),
+        "for content topics",
+        contentTopics,
+        ": ",
+        e
+      );
+      throw e;
```

Comment:
You are doubling up the error usage by printing it now AND throwing it.

Which means that if the consumer of this function also prints the error, then you double up logging entries for no reason.

IMHO, Errors/exception are a one-off usage. You either throw it or print it, not both.

In this instance, I'd try to add context to `e` and throw it. Consumer can then decide to print or discard/handle it.